### DISCUSSION TITLE
Get text from coordinates

### DISCUSSION THREAD

**Discussion by sdezza** [pymupdf/PyMuPDF#2128] [Looking for help] [answered]
Hy guys,

I'm new with this package and what I try to achieve seems basic but I'm struggling.
So, basically the process is as follow:
- Find one specific value
- get the coordinates x0,y0,x1,y1 **(blue)**
- add specific constant 
- get value new value from new coordinate **(red)**

![image](https://user-images.githubusercontent.com/30461090/207885493-f6a6242d-dd9c-457f-8528-c1f30a2876b4.png)

What I have done so far:
```
def find_metadata_value(word, page):
    # Get value from coordinate
    rect = fitz.Rect(word.x0+140, word.y0+1, word.x1+190, word.y1+1)
    word_value = page.get_text("words", clip=rect)
    print( word_value)
```

```
def find_metadata(page, thekeywords):
    # Search keyword from a json file 
      for keyword in keywords['search']:
          words = page.search_for(keyword)
          for word in words:
              word_value = find_metadata_value(word, page)
```

```
document = fitz.open(pdf_files)
document_page_numbers = document.page_count

for page in document:
     find_metadata(page,thekeywords)
```

So even when I'm forcing like for example rect = fitz.Rect(250, 116, 292, 130) It can't find anything.

Am I missing something on the `get_text` fuction?

**Comment by JorjMcKie** [comment]
The `Page.search_for()` method returns a list of **_rectangles_** - one rect for each hit (in the simple case).
If you then do `page.get_textbox(rect)` for one item `rect` of this list, you should be getting the found text.

I do not understand why you modify that rectangle before you try a text extraction. Doing this of course will not let you find anything.
Example:
```python
doc = fitz.open("some.pdf")
page = doc[0]
rlist = page.search_for("pixmap")
for rect in rlist:
    print(page.get_textbox(rect))
```
![grafik](https://user-images.githubusercontent.com/8290722/207898082-885e2494-f4f3-4925-9d40-6031d98d3100.png)

**Reply by JorjMcKie** [reply]
Of course you could have used `page.get_text(clip=rect)` instead of `page.get_textbox()`.

**Reply by sdezza** [reply]
The idea behind modifying the rectangle is, based on the document type and the `search_for` hit,  I know after a certain hit I can go and get a specific value.
In my example above, I need to make a horizontal translation (from blue rectangle to red rectangle).

```
doc = fitz.open("some.pdf")
page = doc[0]
keywords = ["pixmap", "specific", "another"]
for keyword in keywords:
    rlist = page.search_for(keyword)
    ** If keyword "pixmap" is found, go to "(114.5, 115.4, 129.6, 125.5)" and get value from it
    ** If keyword "specific" is found, go to "(253, 116, 292, 126)" and get value from it
    ** etc

```

**Reply by JorjMcKie** [reply]
If you are searching for a string with intervening spaces, and want to have the single words within that string, the approach is much the same. Hereis a search string consisting of a full sentence acorss several pags:
![grafik](https://user-images.githubusercontent.com/8290722/207907042-7d230bad-1e8e-4122-a0a1-f02a62abc40f.png)

**Comment by JorjMcKie** [comment, marked as answer]
> I know after a certain hit I can go and get a specific value.

Ok, got you.
You may have a couple of problems though:
* The actual invoice number may not exactly fit in the same y0/y1 stripe: there always are these uncertainties
* the distance between your keyword rect end and the start of the desired value may be blurry / not always exact

Why dont you do enlarge you hit rect of "invoice number" by some amount. Assuming `rect` is that hit rect, do something like this:
```python
rect.y0 -= 5  # make the stripe higher
rect.y1 += 5
rect.x1 = page.rect.width # go until right page border
words = page.get_text("words", clip=rect, sort=True)  # ensure left-to-right sorting order
# now items 0 and 1 should reflect "Invoice" and "Number" and the rest should be your actual  invoice number.

**Reply by sdezza** [reply]
Much better! didn't get the 'page.rect.width' and `sort` logic 👍 

thank you
